contracts/test/v0.8/ValidatorProxy.test.js

```diff
@@ -0,0 +1,323 @@
+        await expect(validatorProxy.upgradeValidator()).to.be.revertedWith("No proposal");
+      });
+    });
+
+    describe("success", () => {
+      let newValidator;
+      beforeEach(async () => {
+        newValidator = accounts[3].address;
+        await validatorProxy.proposeNewValidator(newValidator);
+      });
+
+      it("should emit an event", async () => {
+        await expect(validatorProxy.upgradeValidator())
+          .to.emit(validatorProxy, "ValidatorUpgraded")
+          .withArgs(validator.address, newValidator);
+      });
+
+      it("should upgrade the addresses", async () => {
+        await validatorProxy.upgradeValidator();
+        const response = await validatorProxy.getValidators();
+        assert.equal(response.current, newValidator);
+        assert.equal(response.hasProposal, false);
+        assert.equal(response.proposed, constants.ZERO_ADDRESS);
+      });
+    });
+  });
+
+  describe("#validate", () => {
+    describe("failure", () => {
+      it("reverts when not called by aggregator or proposed aggregator", async () => {
+        const stranger = accounts[9];
+        await expect(validatorProxy.connect(stranger).validate(99, 88, 77, 66)).to.be.revertedWith(
+          "Not a configured aggregator",
+        );
+      });
+
+      it("reverts when there is no validator set", async () => {
+        const vpf = await ethers.getContractFactory("ValidatorProxy", owner.address);
+        validatorProxy = await vpf.deploy(aggregator.address, constants.ZERO_ADDRESS);
+        await validatorProxy.deployed();
+        await expect(validatorProxy.connect(aggregator).validate(99, 88, 77, 66)).to.be.revertedWith(
+          "No validator set",
+        );
+      });
+    });
+
+    describe("success", () => {
+      describe("from the aggregator", () => {
+        let mockValidator1;
+        beforeEach(async () => {
+          const mvf = await ethers.getContractFactory("MockAggregatorValidator", owner.address);
+          mockValidator1 = await mvf.deploy(1);
+          mockValidator1 = await mockValidator1.deployed();
+          const vpf = await ethers.getContractFactory("ValidatorProxy", owner.address);
+          validatorProxy = await vpf.deploy(aggregator.address, mockValidator1.address);
+          validatorProxy = await validatorProxy.deployed();
+        });
+
+        describe("for a single validator", () => {
+          it("calls validate on the validator", async () => {
+            await expect(validatorProxy.connect(aggregator).validate(200, 300, 400, 500))
+              .to.emit(mockValidator1, "ValidateCalled")
+              .withArgs(1, 200, 300, 400, 500);
+          });
+
+          it("uses a specific amount of gas", async () => {
+            const resp = await validatorProxy.connect(aggregator).validate(200, 300, 400, 500);
+            const receipt = await resp.wait();
+            assert.equal(receipt.gasUsed.toString(), "35371");
```

Comment:
Note I noticed these gasUsed values changed from the original test - I'm assuming thats a function of moving to hardhat/waffle/ethers?

Reply:
Up or down?

Reply:
Went up. This one went from 32510 to 35371 and the other test case went from 40629 to 45318